crates/noirc_evaluator/src/ssa_refactor/ir/dom.rs

```diff
@@ -0,0 +1,423 @@
+//! The dominator tree of a function, represented as a hash map of each reachable block id to its
+//! immediate dominator.
+
+use std::{cmp::Ordering, collections::HashMap};
+
+use super::{basic_block::BasicBlockId, cfg::ControlFlowGraph, post_order::PostOrder};
+
+/// Dominator tree node. We keep one of these per reachable block.
+#[derive(Clone, Default)]
+struct DomNode {
+    /// The block's idx in the control flow graph's reverse post-order
+    reverse_post_order_idx: u32,
+
+    /// The block that immediately dominated that of the node in question.
+    ///
+    /// This will be None for the entry block, which has no immediate dominator.
+    immediate_dominator: Option<BasicBlockId>,
+}
+
+impl DomNode {
+    /// Updates the immediate dominator estimate, returning true if it has changed.
+    ///
+    /// This is used internally as a shorthand during `compute_dominator_tree`.
+    pub(self) fn update_estimate(&mut self, immediate_dominator: BasicBlockId) -> bool {
+        let immediate_dominator = Some(immediate_dominator);
+        if self.immediate_dominator == immediate_dominator {
+            false
+        } else {
+            self.immediate_dominator = immediate_dominator;
+            true
+        }
+    }
+}
+
+/// The dominator tree for a single function.
+pub(crate) struct DominatorTree {
+    /// The nodes of the dominator tree
+    ///
+    /// After dominator tree computation has complete, this will contain a node for every
+    /// reachable block, and no nodes for unreachable blocks.
+    nodes: HashMap<BasicBlockId, DomNode>,
+}
+
+/// Methods for querying the dominator tree.
+impl DominatorTree {
+    /// Is `block_id` reachable from the entry block?
+    pub(crate) fn is_reachable(&self, block_id: BasicBlockId) -> bool {
+        self.nodes.contains_key(&block_id)
+    }
+
+    /// Returns the immediate dominator of `block_id`.
+    ///
+    /// A block is said to *dominate* `block_id` if all control flow paths from the function
+    /// entry to `block_id` must go through the block.
+    ///
+    /// The *immediate dominator* is the dominator that is closest to `block_id`. All other
+    /// dominators also dominate the immediate dominator.
+    ///
+    /// This returns `None` if `block_id` is not reachable from the entry block, or if it is the
+    /// entry block which has no dominators.
+    pub(crate) fn immediate_dominator(&self, block_id: BasicBlockId) -> Option<BasicBlockId> {
+        match self.nodes.get(&block_id) {
+            Some(node) => node.immediate_dominator,
+            _ => None,
+        }
```

Comment:
Nitpick: can we use `.map` here?

Reply:
How would that look? It's an option of an option, and I couldn't find a stdlib flat_map.

Reply:
```suggestion
        self.nodes.get(&block_id).and_then(|node| node.immediate_dominator)
```